### ISSUE TITLE
perf/docs: bucketed sampling and max_seq_len interaction with to_dense padding

### ISSUE THREAD

**Issue by mmcdermott** [mmcdermott/meds-torch-data#104]
## Summary

`MEDSPytorchDataset.collate` calls `JointNestedRaggedTensorDict.to_dense(padding_side=...)`, which pads every sample in the batch to the longest sequence. At high per-batch length variance, this is O(batch × max_len_in_batch) allocation + compute regardless of how short most samples actually are. We don't own the fix (users configure their sampler / `max_seq_len`), but it's an interaction we should document so users hit reasonable defaults.

## Current behavior

[`pytorch_dataset.py:1602-1603`](https://github.com/mmcdermott/meds-torch-data/blob/main/src/meds_torchdata/pytorch_dataset.py#L1602-L1603):

```python
data = JointNestedRaggedTensorDict.vstack([item["dynamic"] for item in batch])
data = data.to_dense(padding_side=self.config.padding_side)
```

`to_dense()` allocates dense tensors sized `(batch, max_seq_len_in_batch, ...)`. A batch of 64 samples where 63 are length 10 and one is length 5000 produces the same tensor shape as 64 samples of length 5000 — most of the storage and GPU transfer is wasted.

## Measured impact

On the toy fixture the effect inverts (shorter skewed data is faster because absolute sizes are smaller):

| Operation | mean | p50 | p95 |
|---|---:|---:|---:|
| `to_dense()` — uniform lengths (4 items) | 50 µs | 50 µs | 54 µs |
| `to_dense()` — skewed (1 item 5×, rest 1×) | 43 µs | 43 µs | 46 µs |

At the fixture scale (lengths 1-5), fixed allocation overhead dominates and the "waste" of padding to 5 from 1 isn't visible. The problem surfaces at much larger variance, e.g. ICU cohorts where some subjects have 10 events and some 50k — the 50k-event subject dictates every batch it lands in. Needs a synthetic skewed-distribution benchmark to measure properly.

## Fix options

All are user-side; MTD itself doesn't need to change its `collate` signature.

### (A) Document `max_seq_len` as the primary knob

The existing `max_seq_len` config caps every sample at `max_seq_len` events in `process_dynamic_data`. If users set `max_seq_len=512` and use a random or to-end sampling strategy, the longest sample in any batch is 512 — bounded memory cost regardless of cohort skew. This is already how things work, but the README doesn't frame it as the "how do I avoid the long-subject-dictates-the-batch problem" answer.

### (B) Bucketed sampling via a user-supplied `BatchSampler`

PyTorch's `torch.utils.data.BatchSampler` + a custom sampler can group samples with similar lengths together so each batch is internally uniform. Users wire their own. MTD can expose the per-subject sequence length (`self.index[i][1]`) as a public accessor so users can build a length-aware sampler without reaching into private attributes.

### (C) Truncate-on-the-fly in collate (not recommended)

Could add a `collate_max_len` kwarg that truncates at collate time. Changes the batch contract in a way users might not expect, and the equivalent is already reachable via `max_seq_len`. Skip.

## Upstream factors

- **nested_ragged_tensors `to_dense`**: if upstream ever ships a "to_dense with per-sample lengths preserved" (ragged tensor / nested tensor output) we could plumb that through to avoid the padding entirely. PyTorch's `torch.nested` is the relevant target but isn't fully API-stable yet. Worth tracking upstream.
- **PyTorch `torch.nested`**: once stable and performant for arbitrary ops, the padded-dense output here becomes optional.

## Benchmark hook

Add a skewed-distribution variant to `benchmark/run.py`:

1. Generate a synthetic cohort where `events_per_subject` follows e.g. a Zipf(1.5) distribution — long tail of subjects with 10×-100× the median event count.
2. Sweep batch_size × `max_seq_len` at the distribution and report mean/p99 batch construction time + peak allocation.
3. Compare against a uniform-length cohort of the same total data volume.

That would catch a regression where MTD accidentally drops the `max_seq_len` truncation (or where `to_dense` changes behavior upstream) — both would show up as a big slowdown on the skewed variant but no change on the uniform one.

## Related

- [#92](https://github.com/mmcdermott/meds-torch-data/issues/92) (closed) — same "skewed distribution blows up a hot path" class of problem, but in `get_task_seq_bounds_and_labels`. This one is in-batch rather than init-time.

**Comment by mmcdermott** [comment]
The real fix here for this is #9 -- just put all the sequences together in one instead of SM or SEM modes at all, then there is no padding, right? It just changes how things are used downstream. I'm not really sure any of the other changes are worth doing here in the interim.

**Comment by mmcdermott** [comment]
Agreed — #9 (offset-based / packed output, no padding at all) is the structural fix. Once sequences are stored as `values` + `cum_lengths` instead of padded-dense, batch skew becomes a non-issue structurally, not "managed by careful sampler config."

The interim options I floated:
- **(A) README doc on `max_seq_len`** — fine to do opportunistically next time the README is touched, but not worth a standalone issue.
- **(B) bucketed sampling helper** — pure bandaid that goes away once #9 lands.
- **(C) truncate-on-collate** — you already flagged I shouldn't do this; agreed.

**Proposal**: close this issue. The one part worth carrying forward is the benchmark hook — a skewed-distribution variant of the benchmark is actually the right **acceptance test** for #9, not an independent ask. Before #9: skewed-batch `collate` time scales with `max_len_in_batch`. After #9: it's flat in batch composition and scales only with total tokens. That's a clean before/after signal.

Want me to:
- (a) Close this and add a comment on #9 capturing the benchmark-variant acceptance test?
- (b) Leave this open but re-title to something like "skewed-batch benchmark variant (acceptance for #9)"?

I'd lean (a) — less issue-tracker clutter, and the benchmark work naturally gets scheduled alongside #9's implementation rather than drifting.

**Comment by mmcdermott** [comment]
Rolled into #9 as the acceptance test — see https://github.com/mmcdermott/meds-torch-data/issues/9#issuecomment-4281143145. Padding skew is a symptom of padded-dense output; #9 (offset-based / packed output) is the structural fix. Closing as superseded.
